hydro_deploy/core/src/ssh.rs

```diff
@@ -89,42 +97,107 @@ impl LaunchedBinary for LaunchedSshBinary {
+            let fold_data = ProgressTracker::leaf("perf script & folding".to_owned(), async move {
+                let mut stderr_lines = FuturesBufReader::new(script_channel.stderr()).lines();
+                let stdout = script_channel.stream(0);
+
+                // Pattern on `()` to make sure no `Result`s are ignored.
+                let ((), fold_data, ()) = tokio::try_join!(
                     async move {
                         // Log stderr.
                         while let Some(Ok(s)) = stderr_lines.next().await {
                             ProgressTracker::println(&format!("[perf stderr] {s}"));
                         }
-                        Ok(()) as Result<_>
+                        Result::<_>::Ok(())
                     },
                     async move {
-                        // Download output via stdout.
-                        // Writing to a file and downloading via `sftp` may be faster as it can download non-sequentially.
-                        Ok(tokio::io::copy(&mut stdout, &mut local_file).await?)
+                        // Download perf output and fold.
+                        tokio::task::spawn_blocking(move || {
+                            let mut fold_data = Vec::new();
+                            fold_er.collapse(
+                                SyncIoBridge::new(TokioBufReader::new(stdout)),
+                                &mut fold_data,
+                            )?;
+                            Ok(fold_data)
+                        })
+                        .await?
                     },
                     async move {
                         // Run command (last!).
-                        Ok(script_channel
+                        script_channel
                             .exec(&format!("perf script --symfs=/ -i {PERF_OUTFILE}"))
-                            .await?)
+                            .await?;
+                        Ok(())
                     },
                 )?;
-            }
+                Result::<_>::Ok(fold_data)
+            })
+            .await?;
+            // Wrap in Arc to allow sharing data across multiple outputs.
+            let fold_data = Arc::new(fold_data);
+            let output_tasks =
+                FuturesUnordered::<Pin<Box<dyn Future<Output = Result<()>> + Send + Sync>>>::new();
+
+            // fold_outfile
+            if let Some(fold_outfile) = perf.fold_outfile.clone() {
+                let fold_data = Arc::clone(&fold_data);
+                output_tasks.push(Box::pin(ProgressTracker::rich_leaf(
+                    format!("write {:?}", fold_outfile),
+                    |set_progress, _| async move {
+                        let mut reader = ProgressReader::new(&**fold_data, set_progress);
+                        let mut writer = tokio::fs::File::create(fold_outfile).await?;
+                        tokio::io::copy_buf(&mut reader, &mut writer).await?;
+                        Ok(())
+                    },
```

Comment:
todo: this seems to complete near-instantly so not really a need to progress track this